osgar/drivers/logzeromq.py

```diff
@@ -76,22 +76,75 @@ def run_output(self):
         socket.close()
         context.term()
 
+    
+    class ReqRepWorker(Thread):
+        def __init__(self, context, bus, endpoint):
+            Thread.__init__ (self)
+            self.context = context
+            self.bus = bus
+
+            self.rossocket = context.socket(zmq.REQ)
+            self.rossocket.RCVTIMEO = int(2000)
+            self.rossocket.connect(endpoint)
+            self.stop_requested = False
+            
+        def stop(self):
+            self.stop_requested = True
+            
+        def run(self):
+            worker = self.context.socket(zmq.REQ)
+            worker.setsockopt(zmq.LINGER, 0)
+            worker.RCVTIMEO = 1000 
+            worker.connect('inproc://reqrepbackend')
+
+            def run_loop():
+                while not self.stop_requested:
+                    try:
+                        worker.send(b"ready")
+                        while True:
+                            try:
+                                data = worker.recv()
+                                break
+                            except zmq.Again:
+                                if self.stop_requested:
+                                    return
+
+                        ident, msg = data.decode('ascii').split('|')
+                        self.rossocket.send_string(msg)
+                        rsp = self.rossocket.recv().decode('ascii')
+                        self.bus.publish('response', [ident, rsp])
+                    except Exception as e:
+                        print("logzeromq thread: %s" % str(e))
+
+            run_loop()
+            worker.close()
+            self.rossocket.close()
+            
     def run_reqrep(self):
         context = zmq.Context()
-        socket = context.socket(zmq.REQ)
-        socket.RCVTIMEO = int(self.timeout * 1000)  # convert to milliseconds
-        socket.connect(self.endpoint)
+        backend = context.socket(zmq.ROUTER)
```

Comment:
I still do not understand the advantage of `ROUTER` against `REQ`. They both are useful, but for Moon project they are just "tunnel" to ROS services, which are blocking request/response ... so why original was not sufficient?

Reply:
We are still at the beginning of the project so this is more of a future-proof building block than a solution to an existing problem. However, the scenarios where async is critical are not hard to imagine. For example, while the code is waiting for the service to come back (for which we do not have any guaranteed performance), the robot keeps moving and keeps receiving more scans and images. We do want to be able to respond to these incoming streams for even while ROS is thinking about the request. 

Second, the original did not work in that after it sent a request, it latched itself on the bus and kept consuming messages until the one it was waiting for appeared. This meant that messages intended for other modules that arrived meanwhile disappeared.

Finally, it's the team's philosophy that "sync calls in robotics are wrong" and that the system design should be async all the time.